From 9b4f53b65bc372514a648b9d23f1990dda3b2a85 Mon Sep 17 00:00:00 2001
From: Kris Chaplin <kris.chaplin@intel.com>
Date: Thu, 29 Apr 2021 09:16:06 +0100
Subject: [PATCH 01/16] net: eth: qse: Add dma_wmb to msgdma prefetcher to
 ensure data ordering

commit 1163217f8f4126bcacc25f4c71d9c6f695f7cb75 from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.114-lts

Use dma_wmb() to make the descriptors are written before setting the go
bit.

Signed-off-by: Kris Chaplin <kris.chaplin@intel.com>
Signed-off-by: Dinh Nguyen <dinh.nguyen@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/altera/altera_msgdma_prefetcher.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/altera/altera_msgdma_prefetcher.c b/drivers/net/ethernet/altera/altera_msgdma_prefetcher.c
index 7eea4537a607..b199992bc7ee 100644
--- a/drivers/net/ethernet/altera/altera_msgdma_prefetcher.c
+++ b/drivers/net/ethernet/altera/altera_msgdma_prefetcher.c
@@ -204,6 +204,12 @@ netdev_tx_t msgdma_pref_tx_buffer(struct altera_dma_private *priv,
 	tx_descs[desc_entry].read_addr_lo = lower_32_bits(buffer->dma_addr);
 	tx_descs[desc_entry].read_addr_hi = upper_32_bits(buffer->dma_addr);
 
+	/*
+	 * Ensure that the high and low address bits of the descriptor are
+	 * written prior to the go bit being set.
+	 */
+	dma_wmb();
+
 	/* set the control bits and set owned by hw */
 	tx_descs[desc_entry].desc_control = (MSGDMA_DESC_CTL_TX_SINGLE
 			| MSGDMA_PREF_DESC_CTL_OWNED_BY_HW);
-- 
2.31.1

